### PR TITLE
Provides a way to register a file watcher relative to a given uri.

### PR DESCRIPTION
**Pull request by heejaechang**
Visual studio code added an ability to watch file/folders outside of workspace folder but LSP didn't expose the ability to server yet. This PR adds that support to LSP.

here are link to spec - https://github.com/microsoft/language-server-protocol/pull/1433

and here are link to vscode API - vscode.workspace.createFileSystemWatcher - https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.d.ts#L11516

### CODE REVIEW

**Comment by heejaechang** [comment]
@dbaeumer can you take a look? and if it looks okay, provide me some guidance for me to finish up the PR?

File: protocol/src/common/protocol.ts [hunk 1]
```diff
@@ -1871,17 +1871,42 @@ export interface DidChangeWatchedFilesRegistrationOptions {
 	watchers: FileSystemWatcher[];
 }
 
+export interface GlobPattern {
```
Comment (dbaeumer) [inline protocol/src/common/protocol.ts:1874]: I would name this RelativePattern and use the type GlobPattern as type GlobPattern = RelativePattern | string;

We might discover other pattern forms in the future.

**Review by dbaeumer — CHANGES_REQUESTED** [review]
We also need to do the following things:

- add a client capability in DidChangeWatchedFilesClientCapabilities. Name proposal is `relativePatternSupport?: boolean`.
- we also need a PR against the spec to add this for 3.17.

File: protocol/src/common/protocol.ts (hunk 1 above)
Comment (heejaechang) [reply protocol/src/common/protocol.ts:1874]: thank you for review @dbaeumer! will make the change!

**Comment by heejaechang** [comment]
> we also need a PR against the spec to add this for 3.17.

what is the spec repo name?

**Comment by heejaechang** [comment]
@dbaeumer can you take a look? I am not sure whether the doc string I added is good enough or I need to add something more

**Review by dbaeumer — APPROVED** [review]

**Comment by dbaeumer** [comment]
I merged it in by hand and did some modifications to it.

**Pull request closed without merging**